[HIP-0026] Plugin tarball support for HTTP and local installers

scottrigby opened the pull request:
Allows installing from a tarball both over HTTP and from local path.

Part of implementing hip-0026. For full context, see:
- https://github.com/helm/helm/issues/31144

Depends on:
- #31146
Please review that PR first. After that PR is merged, will change this PR base branch to `main` and move out of draft.

Checklist:
- [x] This PR contains user facing changes
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

Manually test with:
```sh
% go run ./cmd/helm plugin install https://github.com/scottrigby/h4-example-plugin-types/raw/refs/heads/main/tarballs/example-cli.tgz
Installed plugin: example-cli

# example tarball attached to a GitHub release
% go run ./cmd/helm plugin install https://github.com/scottrigby/h4-example-plugin-types/releases/download/tarball-test-0.2.0/example-legacy-downloader.tgz

# or download a tarball, then install from a local path
% go run ./cmd/helm plugin install PATH/TO/example-cli.tgz
```

scottrigby committed:
Plugin tarball installer support for HTTP (fix) and local (feat)

Signed-off-by: Scott Rigby <scott@r6by.com>

jessesimpson36 commented on internal/plugin/installer/plugin_structure.go:
```diff
@@ -0,0 +1,80 @@
+// detectPluginRoot searches for plugin.yaml in the extracted directory
+// and returns the path to the directory containing it.
+// This handles cases where the tarball contains the plugin in a subdirectory.
+func detectPluginRoot(extractDir string) (string, error) {
```
I'm a little confused by the use-case for `detectPluginRoot` checking one directory deep for a `plugin.yaml` file. Might just be a convenience function.

What it makes me think of is how bitnami charts do something like have a `charts/` directory, maybe someone would have plugins in a `plugins/` directory. But if I ran `helm plugin install`, on `plugins/` then it would be the first directory with a `plugin.yaml` file in it.

Another kind of user might expect a `helm plugin install` to install ALL plugins in that directory.

jessesimpson36 commented on internal/plugin/installer/plugin_structure_test.go:
```diff
@@ -0,0 +1,165 @@
+			},
+			expectRoot:  "my-plugin",
+			expectError: false,
+		},
```
```suggestion
		},
				{
			name: "multiple plugin.yamls in subdirectories",
			setup: func(dir string) error {
				subdir := filepath.Join(dir, "plugins", "plugin1")
				if err := os.MkdirAll(subdir, 0755); err != nil {
					return err
				}
				os.WriteFile(filepath.Join(subdir, "plugin.yaml"), []byte("name: test"), 0644)
				subdir = filepath.Join(dir, "plugins", "plugin2")
				if err := os.MkdirAll(subdir, 0755); err != nil {
					return err
				}
				return os.WriteFile(filepath.Join(subdir, "plugin.yaml"), []byte("name: test2"), 0644)
			},
			expectRoot:  "my-plugin",
			expectError: false,
		},
```

In case you wanted to be more explicit about supporting having multiple folders each with a plugin.yaml, but only getting the first one.

scottrigby replied:
I just think we should align plugin archives with how chart archives are handled.

jessesimpson36 commented on internal/plugin/installer/http_installer_test.go:
```diff
@@ -348,3 +348,255 @@ func TestMediaTypeToExtension(t *testing.T) {
+	tempDir := t.TempDir()
+
+	// Set the umask to default open permissions so we can actually test
+	oldmask := syscall.Umask(0000)
```
I'm inclined to say that this should explicitly open files with the permissions it needs during Open calls.  If you call `syscall.Umask` , I'm wondering if you hit Ctrl-C Ctrl-C in an unfortunate timing, your shell might get stuck creating new files with 777 permissions unless you explicitly close the terminal and restart it. Or for subprocesses of this test case to not know that the umask has changed.

that might be overly paranoid though.

jessesimpson36 reviewed (changes requested):
Ack or respond to my other comments and I'll switch this to approve.

scottrigby committed:
Fix: Removed unsafe umask manipulation in tests

Problem: Tests used syscall.Umask(0000) which could leave your shell creating
files with 777 permissions if interrupted.

Solution: Instead of changing umask, tests now detect the current umask and
calculate expected permissions after it's applied.

Result: Same test coverage, but safe from system-wide side effects.

Co-authored-by: Jesse Simpson <jesse.simpson36@gmail.com>
Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby commented:
pushing an amended commit to trigger build gh workflow

scottrigby committed:
Fix LocalInstaller Path() to strip version from tarball filenames

Override Path() method to use existing stripPluginName function for archives

Signed-off-by: Scott Rigby <scott@r6by.com>

sabre1041 reviewed (approved):
LGTM
